AXI generation docs (#1171)
cucapra/calyx: docs/fud/axi-gen.md

> # Axi generation

Reviewer: ```suggestion
# AXI Interface Generation
```

Calyx currently generates a fairly complex AXI interface that can be daunting
to deal with if confronting for the first time.
The following is an overview of how the generation occurs
and how the output AXI interface behaves as of 2022-9-11.

In general, when `fud` is asked to create an [`.xclbin` file][xclbin] a `kernel.xml`,
`main.sv`, and `toplevel.v` are created as intermediate steps for [xilinx tools][xilinx_tools]
to properly work.

> `main.sv` contains the SystemVerilog needed for our computations to perform

Reviewer 2: We should mention that the file is compiled with slightly different flags, specifically using the `--synthesis` and `-p external` pass pipeline

> correctly. It is implemented as an FSM that derives from the original Calyx program.

Reviewer 2: ```suggestion
correctly. It is implemented as an FSM that derives from the original Calyx program and corresponds to the standard Calyx lowering process.
```

> These files are very hard to read and not really conducive to being examined directly.

Reviewer 2: ```suggestion
```

[`kernel.xml`][kernel_xml] defines register maps and ports of our
toplevel xilinx tools needs. `toplevel.v` wraps our computation kernel and contains
the AXI interface for each memory define in a calyx program. 

## Toplevel

Our toplevel is generated through files in `src/backend/xilinx/`.

### AXI memory controller

> Here, a calyx program is queried and all of its sub components are extracted.

Reviewer 2: This is slightly confusing since only the toplevel component (link to https://docs.calyxir.org/lang/attributes.html?highlight=toplevel#toplevel) queried and only memories marked `@external` (link to https://docs.calyxir.org/lang/attributes.html?highlight=toplevel#external) are turned into AXI buses

Most importantly, memories are extracted and separate AXI interfaces are created for each
one (meaning, there is no shared bus between memories). Each memory has its own

> (single port) bram which writes data taken from an `mi_axi_RDATA` wire where i is the index of

Reviewer: ```suggestion
(single port) BRAM which writes data taken from an `mi_axi_RDATA` wire where `i` is the index of
```

> the memory. Eventually the brams are read and fed into the

Reviewer: ```suggestion
the memory. Eventually the BRAMs are read and fed into the
```

computation kernel of `main.sv`, which outputs results directly into the relevant
memories as defined in the original Calyx program.

> For each memory we dynamically determine necesarry address and data widths and sizes.

Reviewer: necessary (spelling)

> For each memory we dynamically determine necesarry address and data widths and sizes.

Reviewer: Not quite sure what you mean by "dynamically" here. Maybe just say that the pass looks at the cell declaration to determine its size?

However there is always the possiblity that something is hardcoded as a remnant
from previous versions of our AXI generation. If something is hardcoded where it shouldn't

> be please open an [issue][issues].

Reviewer 2: Side comments like these are be nicely handled by moving them into a blockquote using ">" in markdown. That way there is a clear distinction between the content and comments about the content

AXI memory controllers are constructed as (full) [AXI4 managers][signals] that lack a small amount

> of functionality. For example, xPROT signals are not currently supported.

Reviewer 2: Link to documentation that explains `xPROT` signal.

> Additionally, things like bursting are not currently supported, but should be

Reviewer 2: Link to `bursting`

easy to implement due to the existing infrastructure and generation.

A list of current signals that are hardcoded follows:

> * `xLEN` is set to 0, correlating to aburt length of 1.

Reviewer: ```suggestion
* `xLEN` is set to 0, correlating to a burst length of 1.
```

> * `xBURST` is set to 01, correlating to INCR type of bursts.

Reviewer: ```suggestion
* `xBURST` is set to 01, corresponding to INCR type of bursts.
```
